actionpack/lib/action_dispatch/middleware/actionable_exceptions.rb

```diff
@@ -16,24 +16,54 @@ def call(env)
       request = ActionDispatch::Request.new(env)
       return @app.call(env) unless actionable_request?(request)
 
-      ActiveSupport::ActionableError.dispatch(request.params[:error].to_s.safe_constantize, request.params[:action])
+      actionable_error = request.params[:error].to_s.safe_constantize
+      action = request.params[:action]
 
-      redirect_to request.params[:location]
+      rendering_response(request, actionable_error) do
+        ActiveSupport::ActionableError.dispatch(actionable_error, action)
+      end
+    rescue Exception => error
+      ActiveSupport::ActionableError.raise_if_triggered_by(error)
+      raise
     end
 
     private
       def actionable_request?(request)
         request.show_exceptions? && request.post? && request.path == endpoint
       end
 
-      def redirect_to(location)
-        body = "<html><body>You are being <a href=\"#{ERB::Util.unwrapped_html_escape(location)}\">redirected</a>.</body></html>"
+      def rendering_response(request, error)
+        body = capturing_standard_streams { yield }
 
-        [302, {
-          "Content-Type" => "text/html; charset=#{Response.default_charset}",
+        [200, {
+          "Content-Type" => "text/plain; charset=#{Response.default_charset}",
           "Content-Length" => body.bytesize.to_s,
-          "Location" => location,
         }, [body]]
       end
+
+      def capturing_standard_streams
+        capture :stdout do
+          capture :stderr do
+            yield
+          end
+        end
+      end
+
+      def capture(stream)
+        stream = stream.to_s
+        captured_stream = Tempfile.new(stream)
+        stream_io = eval("$#{stream}")
+        origin_stream = stream_io.dup
+        stream_io.reopen(captured_stream)
+
+        yield
```

Comment:
How are we actually capturing `$stderr` since the result from yield here isn't flowing up to `capture :stdout` far as I can see?

Don't we actually want stdout and stderr to share the same Tempfile io? Aren't the messages out of order otherwise? And does it have to be a Tempfile or is a plain `StringIO.new` good enough?

Would this work?

```ruby
def capture_ios
  ios = [ $stdout, $stderr ]
  $stdout = $stderr = shared_io = StringIO.new

  yield

  shared_io.string
ensure
  $stdout, $stderr = ios
end
```

Reply:
The `capture_ios` will capture output issued from the current process (say we invoke a method in the action and it `puts`), but won't caputre output from child processes like the ones created during `system` calls. We call `Rails::Command.invoke` that invokes rake tasks that usually start other processes. You have to `reopen` the streams for capturing the chile process output as the assignment does not propagate the streams to the children, but keeps the "original" ones.

I have spent more time on a capturing implementation that is good with both child capture and current process output handling.